lib/index.js

```diff
@@ -17,40 +30,602 @@ const PERMISSIONS = bedrock.config.permission.permissions;
+    projection.id = 1;
+  }
+  if(email) {
+    query['account.email'] = email;
+    projection['account.email'] = 1;
+  }
+  return !!await database.collections.account.findOne(query, projection);
+});
+
+/**
+ * Inserts a new Account. The Account must contain `id`.
+ *
+ * @param actor the actor or capabilities for performing the action.
+ * @param account the Account containing at least the minimum required data.
+ * @param [meta] optional meta information to include.
+ *
+ * @return a Promise that resolves to the database account record.
+ */
+api.insert = brCallbackify(async ({actor, account, meta}) => {
+  assert.object(account, 'account');
+  assert.string(account.id, 'account.id');
+  assert.optionalString(account.email, 'account.email');
+  assert.optionalString(account.phoneNumber, 'account.phoneNumber');
+
+  meta = Object.assign({}, meta, {status: 'active'});
+
+  await _checkPermission({
+    actor,
+    permission: PERMISSIONS.ACCOUNT_INSERT,
+    resource: [account]
+  });
+
+  // emit `insertEvent` with clone of `account`
+  account = bedrock.util.clone(account);
+  const eventData = {
+    actor,
+    account,
+    meta,
+    // data to pass to `postInsert`, but do not insert into database
+    postInsert: {
+      /* <module-name>: <module-specific data> */
+    }
+  };
+  await bedrock.events.emit('bedrock-account.insert', eventData);
+
+  // replay assertions post event emission
+  assert.object(account, 'account');
+  assert.string(account.id, 'account.id');
+  assert.string(meta.status, 'meta.status');
+  assert.optionalString(account.email, 'account.email');
+  assert.optionalString(account.phoneNumber, 'account.phoneNumber');
+
+  // generate resource role resources
+  const roles = meta.sysResourceRole = meta.sysResourceRole || [];
+  for(let i = 0; i < roles.length; ++i) {
+    const role = roles[i];
+    if(role.generateResource === 'id') {
+      roles[i] = api.generateResource({role, id: account.id});
+    } else if(role.generateResource) {
+      // unknown generation directive
+      throw new BedrockError(
+        'Could not create Account; unknown ResourceRole rule.',
+        'InvalidResourceRole', {sysResourceRole: role});
+    }
+  }
+
+  // validate resource roles (ensure actor is permitted to delegate the
+  // roles specified in the account meta)
+  if(meta.sysResourceRole.length > 0) {
+    await validateCapabilityDelegation(
+      {actor, resourceRoles: meta.sysResourceRole});
+  }
+
+  logger.info('inserting account', {account});
+
+  // insert the account and get the updated record
+  const now = Date.now();
+  meta.created = now;
+  meta.updated = now;
+  meta.sequence = 0;
+  let record = {
+    id: database.hash(account.id),
+    meta,
+    account
+  };
+  try {
+    const result = await database.collections.account.insert(
+      record, database.writeOptions);
+    record = result.ops[0];
+  } catch(e) {
+    if(!database.isDuplicateError(e)) {
+      throw e;
+    }
+    throw new BedrockError(
```

Comment:
It's good to include the 3rd argument (cause) here `e`.  Especially nice for `DuplicateError` because the cause will contain the information about the unique index.  Sorry if the example I provided for this did not do that.